Open Judge System/Workers/OJS.Workers/LocalWorkerServiceBase.cs

```diff
@@ -0,0 +1,141 @@
+﻿namespace OJS.Workers
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Reflection;
+    using System.ServiceProcess;
+    using System.Threading;
+
+    using log4net;
+
+    using OJS.Common;
+    using OJS.Workers.Common;
+    using OJS.Workers.Jobs;
+
+    public class LocalWorkerServiceBase<TSubmission> : ServiceBase
+    {
+        private readonly ICollection<Thread> threads;
+        private readonly ICollection<IJob> jobs;
+
+        protected LocalWorkerServiceBase()
+        {
+            var loggerAssembly = Assembly.GetEntryAssembly() ?? Assembly.GetCallingAssembly();
+
+            this.Logger = LogManager.GetLogger(loggerAssembly, Constants.LocalWorkerServiceLogName);
+
+            this.threads = new List<Thread>();
+            this.jobs = new List<IJob>();
+        }
+
+        protected ILog Logger { get; }
+
+        protected IDependencyContainer DependencyContainer { get; private set; }
+
+        protected override void OnStart(string[] args)
+        {
+            this.Logger.Info("LocalWorkerService starting...");
+
+            this.DependencyContainer = this.GetDependencyContainer();
+
+            this.SpawnJobsAndThreads(this.jobs, this.threads, new ConcurrentQueue<TSubmission>());
+
+            this.BeforeStartingThreads();
+
+            this.StartThreads(this.threads);
+
+            this.Logger.Info("LocalWorkerService started.");
+        }
+
+        protected override void OnStop()
+        {
+            this.Logger.Info("LocalWorkerService stopping...");
+
+            this.StopJobs(this.jobs);
+
+            Thread.Sleep(this.TimeBeforeAbortingThreadsInMilliseconds);
+
+            this.StopThreads(this.threads);
+
+            this.Logger.Info("LocalWorkerService stopped.");
+        }
+
+        protected virtual void BeforeStartingThreads()
+        {
+            this.CreateExecutionStrategiesWorkingDirectory();
+        }
+
+        protected virtual IDependencyContainer GetDependencyContainer() =>
```

Comment:
Is there some limititation for this to be `abstract`?

Reply:
Yes, it inherits `ServiceBase` class and that's why it cannot be abstract